src/util/sighash.rs

```diff
@@ -0,0 +1,904 @@
+        let mut enc = SigHash::engine();
+        self.legacy_encode_signing_data_to(&mut enc, input_index, script_pubkey, sighash_type)?;
+        Ok(SigHash::from_engine(enc))
+    }
+
+    fn common_cache(&mut self) -> &CommonCache {
+        if self.common_cache.is_none() {
+            let mut enc_prevouts = sha256::Hash::engine();
+            let mut enc_sequences = sha256::Hash::engine();
+            for txin in self.tx.input.iter() {
+                txin.previous_output
+                    .consensus_encode(&mut enc_prevouts)
+                    .unwrap();
+                txin.sequence.consensus_encode(&mut enc_sequences).unwrap();
+            }
+            let cache = CommonCache {
+                prevouts: sha256::Hash::from_engine(enc_prevouts),
+                sequences: sha256::Hash::from_engine(enc_sequences),
+                outputs: {
+                    let mut enc = sha256::Hash::engine();
+                    for txout in self.tx.output.iter() {
+                        txout.consensus_encode(&mut enc).unwrap();
+                    }
+                    sha256::Hash::from_engine(enc)
+                },
+            };
+            self.common_cache = Some(cache);
+        }
+        self.common_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
+    }
+
+    fn segwit_cache(&mut self) -> &SegwitCache {
+        if self.segwit_cache.is_none() {
+            let cache = SegwitCache {
+                prevouts: sha256d::Hash::from_inner(
+                    sha256::Hash::hash(&self.common_cache().prevouts).into_inner(),
+                ),
+                sequences: sha256d::Hash::from_inner(
+                    sha256::Hash::hash(&self.common_cache().sequences).into_inner(),
+                ),
+                outputs: sha256d::Hash::from_inner(
+                    sha256::Hash::hash(&self.common_cache().outputs).into_inner(),
+                ),
+            };
+            self.segwit_cache = Some(cache);
+        }
+        self.segwit_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
+    }
+
+    fn taproot_cache(&mut self, prevouts: &[TxOut]) -> &TaprootCache {
+        if self.taproot_cache.is_none() {
+            let mut enc_amounts = sha256::Hash::engine();
+            let mut enc_script_pubkeys = sha256::Hash::engine();
+            for prevout in prevouts {
+                prevout.value.consensus_encode(&mut enc_amounts).unwrap();
+                prevout
+                    .script_pubkey
+                    .consensus_encode(&mut enc_script_pubkeys)
+                    .unwrap();
+            }
+            let cache = TaprootCache {
+                amounts: sha256::Hash::from_engine(enc_amounts),
+                script_pubkeys: sha256::Hash::from_engine(enc_script_pubkeys),
+            };
+            self.taproot_cache = Some(cache);
+        }
+        self.taproot_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
```

Comment:
// ... and we produced `Some()` at the last line inside `if`.

Reply:
No need to change anymore since I included your PR https://github.com/RCasatta/rust-bitcoin/pull/3